### DISCUSSION TITLE
gstreamer pipeline unable to find tensorflow version installed in the Ubuntu system

### DISCUSSION THREAD

**Discussion by aks473** [nnstreamer/nnstreamer#3887] [General]
It might be very trivial but i need help. I'm trying to run nnstreamer for the first time , prepared a fresh Ubuntu 18.04, installed gstreamer by following instructions from Gstreamer official site. Then installed tensorflow 1.13.1. Followed instruction for installing nnstreamer using PPA and nnstreamer examples also using PPA. 
When i tried to run gstreamer_objectDetection.sh sample , i'm getting an error --> Cannot identify the neural network framework , tensorflow. Even though i've installed it. and i think because of that i'm seeing another error could not link tensorconvertor0 to tensorfilter0

**Comment by myungjoo** [comment]
You need to let others recreate the issue you are experiencing.
Please show the shortest and simplest code and procedures to recreate your issue.

Please run the utility, ```nnstreamer-check```, and show the result as well as your pipeline and its error/warning messages.

**Reply by aks473** [reply]
<img width="964" alt="image" src="https://user-images.githubusercontent.com/6905286/185561198-843b8333-e4ea-44ca-970a-dad6ba2b79f0.png">
<img width="1249" alt="image" src="https://user-images.githubusercontent.com/6905286/185561438-1f71469d-a2b1-4336-91d5-7bb9f74f5c63.png">

Followed the steps mentioned in here for installing nnstreamer and samples. As you could see in the screenshot i'm still getting error while running the gst-launch-object-detection-tf.sh.
I also have tensorflow=1.13.1 and gstreamer installed in the system.

Basically i have setup a fresh Ubuntu 18.04 with 
1)  installed tensorflow 1.13 on it
2) installed gstreamer by following instructions at 
<img width="779" alt="image" src="https://user-images.githubusercontent.com/6905286/185563367-f37660b3-e9ba-41e8-969c-0fcab04ad2cd.png">
3) installed nnstreamer (https://nnstreamer.ai/#get-started)

4) installed nnstreamer examples

I'm definitely missing something here, must be very basic
screenshot of printenv
<img width="674" alt="image" src="https://user-images.githubusercontent.com/6905286/185563992-20557891-c727-4130-85cb-94665e83f973.png">

**Comment by myungjoo** [comment]
We need to check
1. is nnstreamer properly installed?
    - install nnstreamer-util from PPA. It will show related info w/ ```nnstreamer-check```
2. is nnstreamer-tensorflow property built and installed?
    - If you want to use tensorflow of your own, not from nnstreamer PPA, you may need to rebuild nnstreamer-tensorflow depending on the versions and additional features/patches of your tensorflow. Try tensorflow and nnstreamer-tensorflow from nnstreamer ppa first. 
3. what is your pipeline? (shortest and simplest that recreates your issue)

**Reply by aks473** [reply]
Thanks a ton for the propmt response.

Output of nnstreamer-check
<img width="737" alt="image" src="https://user-images.githubusercontent.com/6905286/185569676-75e03203-b91d-4992-a6fb-4e2fbbac9a4c.png">
I've just started exploring nnstreamer and i wanted it to run with the instructions provided on the site. I have no need for having my own tensorflow version , i thought nnstreamer might need it to run the samples. That's the reason i had installed gstreamer and tensorflow prior to installing nnstreamer. 

Our product is using gstreamer/DLStreamer and Openvino , 
so i'm trying the pipelines available under nnstreamer samples like object detection to get a hang of it and decide on how we could leverage nnstreamer
<img width="673" alt="image" src="https://user-images.githubusercontent.com/6905286/185570889-529efefa-d054-4437-ac03-778ded99b04f.png">

**Reply by aks473** [reply]
I had reinstalled nnstreamer, it's plugins , nnstreamer-example without  on a fresh Ubuntu 18.04 but still facing same issue of couldn't identify the give NN Framework while running gst-launch-object-detection-tf.sh . Didn't install any tensorflow version separately on the system.
![image](https://user-images.githubusercontent.com/6905286/185614948-79254a65-8ab4-41bc-be2e-f4a98085bd8b.png)

![image](https://user-images.githubusercontent.com/6905286/185615364-b5edd523-cba2-43e9-a3cf-4538b8959807.png)

**Reply by myungjoo** [reply]
You don't have nnstreamer-tensorflow installed. With tensorflow subplugin, nnstreamer-check should look like:

```
...
[Filter]
  tensorflow
    - No custom property found
  python3
    - ${GENERAL_STRING}: There is no key-value pair defined by python3 subplugin. Provide arguments for the given python3 script.
  tensorflow1-lite
    - NumThreads: Number of threads. Set 0 for default behaviors.
    - Delegate: TF-Lite delegation options: {'NNAPI', 'GPU', 'XNNPACK', 'External'}. Do not specify to disable delegation.
    - ExtDelegateLib: Path to external delegate shared library
    - ExtDelegateKeyVal: key/values pairs optional parameters for delegate. Format ExtDelegateKeyVal=key1#value1;key2#value2...
  cpp
    - No custom property found
...
```

So, 
1. install tensorflow and nnstreamer-tensorflow from PPA or (the fastest method. if you have installed nnstreamer via ppa; do ```$ sudo apt install nnstreamer-tensorflow```)
2. build nnstreamer-tensorflow in your system with tensorflow & its developmental files ready. (if you are familiar with build&install from C/C++ sources in Linux)

When you build(rebuild) nnstreamer w/ meson, add ```-Dtf-support=enabled``` (refer to meson_options.txt) to ensure that meson fails if it cannot find a proper installation of tensorflow; otherwise, it will simply skip building nnstreamer-tensorflow.
